spring-batch-core/src/main/java/org/springframework/batch/core/repository/support/SimpleJobRepository.java

```diff
@@ -152,13 +153,15 @@ public JobExecution createJobExecution(String jobName, JobParameters jobParamete
 							+ "so it may be dangerous to proceed. Manual intervention is probably necessary.");
 				}
 				Collection<JobParameter<?>> allJobParameters = execution.getJobParameters().getParameters().values();
-				long identifyingJobParametersCount = allJobParameters.stream().filter(JobParameter::isIdentifying)
-						.count();
+				Collection<JobParameter<?>> identifyingJobParameters = allJobParameters.stream()
+						.filter(JobParameter::isIdentifying).toList();
+				long identifyingJobParametersCount = identifyingJobParameters.size();
 				if (identifyingJobParametersCount > 0
 						&& (status == BatchStatus.COMPLETED || status == BatchStatus.ABANDONED)) {
 					throw new JobInstanceAlreadyCompleteException(
-							"A job instance already exists and is complete for parameters=" + jobParameters
-									+ ".  If you want to run this job again, change the parameters.");
+							"A job instance already exists and is complete for identifying parameters="
+									+ identifyingJobParameters + ".  If you want to run this job again, "
```

Comment:
This will slightly modify the exception's message as it now uses `Collection#toString` rather than `JobParameters#toString`. I will update this on merge by wrapping the identifying job parameters in a `JobParameters` instance.